### ISSUE TITLE
Question about the "loss_of_one_batch_tbptt" function when training.

### ISSUE THREAD

**Issue by runnerhdh** [CUT3R/CUT3R#108]
Thank you again for this excellent open-source work.

As you mentioned in your train.md, when training from scratch, there are four stages: the first three stages have 4 views each, and the fourth stage has 32 views. Why isn't there a 16-view stage in the middle, such as the third stage?

Also, in `linear_224_fixed_16`, `long context=false`, meaning no TBPTT operations, the state gradient can be backpropagated from the last to the first. When the number of views is expanded to 32 and beyond, TBPTT is used, and only the gradients of the last four chunks can be backpropagated, with the state gradients detached between each chunk. I guess this means the state gradients are continuous within every four images. What is the purpose of detaching the state gradients between chunks? Why not allow the state gradients to be continuous within the last 16 frames, like in `linear_224_fixed_16`? Would it be more helpful to allow the state gradient to flow between more frames to improve the capabilities of long sequences?

Looking forward to your reply, thank you very much :)

**Comment by Forrest-110** [comment]
Hi @runnerhdh! We use 512 resolution in stages 3 and 4, and unfortunately, 16-view at 512 don’t fit on our GPUs. That said, training a 16-view 512 stage in between would potentially be very helpful.

In our experiments at 512 resolution, training dynamics are stable with 4 images per sequence, but convergence becomes much more difficult with 16 views. More generally, we’re not fully confident about the gradient behavior for long sequences at 512 with DPT head. For that reason, we chose a conservative setup and only backpropagate through 4 views.

**Comment by runnerhdh** [comment]
> Hi [@runnerhdh](https://github.com/runnerhdh)! We use 512 resolution in stages 3 and 4, and unfortunately, 16-view at 512 don’t fit on our GPUs. That said, training a 16-view 512 stage in between would potentially be very helpful.
> 
> In our experiments at 512 resolution, training dynamics are stable with 4 images per sequence, but convergence becomes much more difficult with 16 views. More generally, we’re not fully confident about the gradient behavior for long sequences at 512 with DPT head. For that reason, we chose a conservative setup and only backpropagate through 4 views.

Thanks for your replying:)
I still have some questions about some settings. In stage 4 with  fixed 32 view length and multi resolution, only the last 16 frames were trainable. The difference between this method and "16-view at 512 don’t fit on our GPUs." is that the former state performed 4-frame detachment. Do you mean that if chunk detachment isn't done, there won't be enough GPU memory for training? Also, the number of views jumped from 4 to 32 in stages 3 and 4. The latter requires relatively stable state across 16 frames (because gradients aren't calculated), but stage 3 only calculated 4 views. Why is the jump so large? Did you perform any other operations to ensure the stability of the 16-frame state?
Looking forward to your reply, thank you very much!!!!

**Comment by Forrest-110** [comment]
> In stage 4 with fixed 32 view length and multi resolution, only the last 16 frames were trainable. 

I believe the chunk size is always 4, as indicated [here](https://github.com/CUT3R/CUT3R/blob/8bc15dc92a6d7fd92920b4ec81540d3dec7d3ecf/src/train.py#L438).

> Do you mean that if chunk detachment isn't done, there won't be enough GPU memory for training?

Yes.

>  The latter requires relatively stable state across 16 frames.

Since the chunk_size is 4, stability over 4 frames should be sufficient.
